src/Microsoft.AspNet.OData.Shared/Query/SkipTokenHandler.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using Microsoft.AspNet.OData.Formatter.Serialization;
+
+namespace Microsoft.AspNet.OData.Query
+{
+    /// <summary>
+    /// Represents how NextLink for paging is generated.
+    /// </summary>
+    public abstract class SkipTokenHandler
+    {
+        /// <summary>
+        /// Raw SkipToken Value 
+        /// </summary>
+        public abstract string Value { get; set; }
+
+        /// <summary>
+        /// Indicates whether generating a nextlink for delta feed is supported.
+        /// </summary>
+        public bool IsDeltaFeedSupported { get; set; }
+
+        /// <summary>
+        /// Gets and sets the given <see cref="ODataQueryContext"/>.
+        /// </summary>
+        public ODataQueryContext Context { get; set; }
+
+        /// <summary>
+        /// Apply the $skiptoken query to the given IQueryable.
+        /// </summary>
+        /// <param name="query">The original <see cref="IQueryable"/>.</param>
+        /// <param name="skipTokenQueryOption">The query option that contains all the relevant information for applying skiptoken.</param>
+        /// <returns>The new <see cref="IQueryable"/> after the skiptoken query has been applied to.</returns>
+        public abstract IQueryable<T> ApplyTo<T>(IQueryable<T> query, SkipTokenQueryOption skipTokenQueryOption);
+
+        /// <summary>
+        /// Apply the $skiptoken query to the given IQueryable.
+        /// </summary>
+        /// <param name="query">The original <see cref="IQueryable"/>.</param>
+        /// <param name="skipTokenQueryOption">The query option that contains all the relevant information for applying skiptoken.</param>
+        /// <returns>The new <see cref="IQueryable"/> after the skiptoken query has been applied to.</returns>
+        public abstract IQueryable ApplyTo(IQueryable query, SkipTokenQueryOption skipTokenQueryOption);
+
+        /// <summary>
+        /// Returns the URI for NextPageLink
+        /// </summary>
+        /// <param name="baseUri">BaseUri for nextlink.</param>
+        /// <param name="pageSize">Maximum number of records in the set of partial results for a resource.</param>
+        /// <param name="instance">Instance based on which SkipToken value will be generated.</param>
+        /// <param name="context">Serializer context</param>
+        /// <returns></returns>
+        public abstract Uri GenerateNextPageLink(Uri baseUri, int pageSize, Object instance, ODataSerializerContext context);
```

Comment:
So I tried to implement a custom SkipTokenHandler. I was kind of blocked by the fact that only the GenerateNextPageLink method is extensible, and not the method GenerateSkipTokenValue (which was extensible in the previous PR).
In addition to not being extensible, most of the functions that can be used to generate a link are internal, meaning that I cannot really recreate a link with my custom value without duplicating a lot of code.

So my suggestion is to extract the method `GenerateSkipTokenValue` into a separate class that implements a `ISkipTokenValueGenerator` to that it can we can inject a custom implementation.

Reply:
@Galad I was unable to pick this up in the meanwhile. While I like your suggestion, I am not leaning towards having DI support just for the value of the skiptoken. I have made the existing default class be more extensible and that can be used to generate the nextlink.